### PR TITLE
feat(self-trigger): add backend pagination for the table

### PR DESCRIPTION
**Pull request by simcha90**
This PR add backend pagination for `plugin-chart-table`.
Requires: https://github.com/apache/superset/pull/13137
As part of this request done next changes:
1. Getting ownState of table chart from dashboard with `{ pageSize, currentPage }`
2. It's update `buildQuery` of this plugin with adding `row_offset` and `row_limit`:
Because BE still not support returning of total rows available for current request, we do request for number items per page `+1` to know if there is available one more page. For same reason we use only `previous/next` button in pagination
3. Because we needed update definition of `setExtraFormData` hook, we also updated it's name to more relevant for it's action: `setDataMask`
4. Added `custom_form_data` to `extraFormData` object to pass custom props for build necessary request to the server in `buildQuery`

https://user-images.githubusercontent.com/56388545/107976853-e4544a80-6fc2-11eb-9eb6-988b07a755f7.mov

### CODE REVIEW

**Review by ktmud — COMMENTED** [review]
Thanks for working on this highly-demanded feature! A lot of users would be grateful!

Since the queryContext can now run multiple queries, I'm wondering how difficult would it be to let `buildQuery` add an additional `SELECT count(*) FROM {...subquery...}` query, which may be implemented by adding another [`ChartDataResultType`](https://github.com/apache/superset/blob/ac73991913c596a0b2b37aaf96083c227088dad6/superset/common/query_context.py#L167)?

**Comment by amitmiran137** [comment]
It was difficult to achieve that right now 
I think we could get to that together with the row totals feature

**Comment by villebro** [comment]
> Since the queryContext can now run multiple queries, I'm wondering how difficult would it be to let `buildQuery` add an additional `SELECT count(*) FROM {...subquery...}` query, which may be implemented by adding another [`ChartDataResultType`](https://github.com/apache/superset/blob/ac73991913c596a0b2b37aaf96083c227088dad6/superset/common/query_context.py#L167)?

This is definitely on the short term roadmap. I didn't look to closely into how much work this will be, but I believe we need to do some minor refactoring to achieve this. I had been planning on adding some optional props from `QueryContext` to `QueryObject` to make it possible to introduce support for querying data from multiple different datasources in a single `QueryContext`, and I believe the same could be relevant for `result_type` (=mixing `RESULTS` with `SAMPLES` or the forthcoming `ROWCOUNT`).

Anyway, let's defer this feature until the backend feature is merged, as there's no clean ANSI SQL way to achieve multi-column COUNT DISTINCTs without a subquery.

File: packages/superset-ui-core/src/chart/types/Base.ts [hunk 1]
```diff
@@ -7,13 +7,24 @@ export enum Behavior {
...
+      currentState: { value: any; [key: string]: any };
+    };
+    cross?: {
+      extraFormData: ExtraFormData;
+      currentState: { [key: string]: any };
+    };
+    own?: {
+      extraFormData: ExtraFormData;
+      currentState: { [key: string]: any };
+    };
```
Comment (villebro) [inline packages/superset-ui-core/src/chart/types/Base.ts:27]: When seeing this like this, I think new developers will be confused by what `native`, `cross` and `own` mean. Not sure I have better proposals, but at the risk of being misleading if the scope of this functionality is expanded more, I think `nativeFilters`, `crossFilters` and `ownFilters` could be slightly more descriptive. Alternatively we might want to add a comment per prop to help understand what the context is.

File: packages/superset-ui-core/src/query/types/QueryFormData.ts [hunk 2]
```diff
@@ -100,6 +101,7 @@ export type QueryFormExtraFilter = {
 export type ExtraFormData = {
   append_form_data?: Partial<QueryObject>;
   override_form_data?: Partial<QueryObject>;
+  custom_form_data?: JsonObject;
```
Comment (villebro) [inline packages/superset-ui-core/src/query/types/QueryFormData.ts:104]: A comment here could also be helpful (also add to the other two)

File: packages/superset-ui-core/src/query/processExtraFormData.ts [hunk 3]
```diff
@@ -18,7 +18,14 @@
...
 import { QueryObject } from './types/Query';
 
-const ALLOWED_OVERRIDES = ['time_grain_sqla', 'time_range', 'since', 'until'];
+const ALLOWED_OVERRIDES = [
+  'time_grain_sqla',
+  'time_range',
+  'since',
+  'until',
+  'row_offset',
+  'row_limit',
```
Comment (villebro) [inline packages/superset-ui-core/src/query/processExtraFormData.ts:27]: These might no longer be needed

File: plugins/plugin-chart-table/src/DataTable/utils/externalAPIs.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+import { SetDataMaskHook } from '@superset-ui/core';
+
+export const updateExternalFormData = (
+  setDataMask: SetDataMaskHook,
+  pageNumber: number,
+  pageSize: number,
+) =>
+  setDataMask({
+    // @ts-ignore
```
Comment (villebro) [inline plugins/plugin-chart-table/src/DataTable/utils/externalAPIs.ts:28]: is the ignore needed?

File: plugins/plugin-chart-table/src/DataTable/DataTable.tsx [hunk 5]
```diff
@@ -29,10 +36,14 @@ import {
   Row,
 } from 'react-table';
 import { matchSorter, rankings } from 'match-sorter';
+import { SetDataMaskHook } from '@superset-ui/core';
 import GlobalFilter, { GlobalFilterProps } from './components/GlobalFilter';
 import SelectPageSize, { SelectPageSizeProps, SizeOption } from './components/SelectPageSize';
 import SimplePagination from './components/Pagination';
 import useSticky from './hooks/useSticky';
+import { updateExternalFormData } from './utils/externalAPIs';
+import BackendPagination from './components/BackendPagination';
```
Comment (villebro) [inline plugins/plugin-chart-table/src/DataTable/DataTable.tsx:45]: `ServerPagination` or `DatabasePagination` might be a more appropriate name

**Comment by ktmud** [comment]
Another question, how does this interact with percent metrics? Are we OK with percent metrics only be calculated based on limited rows? It'd be somewhat pointless to calculate distribution only on one page of data (10 ~ 200 rows) when the total number of records may be millions.

**Comment by villebro** [comment]
> Another question, how does this interact with percent metrics? Are we OK with percent metrics only be calculated based on limited rows? It'd be somewhat pointless to calculate distribution only on one page of data (10 ~ 200 rows) when the total number of records may be millions.

This will need to be addressed in a forthcoming PR - once we have the second query for totals we can calculate the percent mertrics in the frontend, bypassing the need for the backend postprocessing and always getting percent distributions for the full dataset.

File: plugins/plugin-chart-table/src/DataTable/DataTable.tsx [hunk 6]
```diff
@@ -109,7 +129,7 @@ export default function DataTable<D extends object>({
     }
     return undefined;
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [initialHeight, initialWidth, wrapperRef, hasPagination, hasGlobalControl]);
+  }, [initialHeight, initialWidth, wrapperRef, hasPagination, hasGlobalControl, showBENextButton]);
```
Comment (villebro) [inline plugins/plugin-chart-table/src/DataTable/DataTable.tsx:132]: showNextButton

File: plugins/plugin-chart-table/src/DataTable/DataTable.tsx [hunk 7]
```diff
@@ -29,10 +36,14 @@ import {
...
 } from 'react-table';
 import { matchSorter, rankings } from 'match-sorter';
+import { SetDataMaskHook } from '@superset-ui/core';
 import GlobalFilter, { GlobalFilterProps } from './components/GlobalFilter';
 import SelectPageSize, { SelectPageSizeProps, SizeOption } from './components/SelectPageSize';
 import SimplePagination from './components/Pagination';
 import useSticky from './hooks/useSticky';
+import { updateExternalFormData } from './utils/externalAPIs';
+import ServerPagination from './components/ServerPagination';
+import { BackendPage } from '../types';
```
Comment (villebro) [inline plugins/plugin-chart-table/src/DataTable/DataTable.tsx:46]: For consistency, this should probably  be `ServerPage`

File: plugins/plugin-chart-table/src/DataTable/DataTable.tsx [hunk 8]
```diff
@@ -43,9 +54,13 @@ export interface DataTableProps<D extends object> extends TableOptions<D> {
   hooks?: PluginHook<D>[]; // any additional hooks
   width?: string | number;
   height?: string | number;
+  backendPagination?: boolean;
```
Comment (villebro) [inline plugins/plugin-chart-table/src/DataTable/DataTable.tsx:57]: ..serverPagination

File: plugins/plugin-chart-table/src/DataTable/DataTable.tsx [hunk 9]
```diff
@@ -66,9 +82,12 @@ export default function DataTable<D extends object>({
   maxPageItemCount = 9,
   sticky: doSticky,
   searchInput = true,
+  setDataMask,
+  showBENextButton,
   selectPageSize,
   noResults: noResultsText = 'No data found',
   hooks,
+  backendPagination,
```
Comment (villebro) [inline plugins/plugin-chart-table/src/DataTable/DataTable.tsx:90]: ..serverPagination

File: plugins/plugin-chart-table/src/DataTable/DataTable.tsx [hunk 10]
```diff
@@ -43,9 +54,13 @@ export interface DataTableProps<D extends object> extends TableOptions<D> {
   hooks?: PluginHook<D>[]; // any additional hooks
   width?: string | number;
   height?: string | number;
+  backendPagination?: boolean;
+  setDataMask: SetDataMaskHook;
+  currentPage?: number;
   pageSize?: number;
   noResults?: string | ((filterString: string) => ReactNode);
   sticky?: boolean;
+  showBENextButton: boolean;
```
Comment (villebro) [inline plugins/plugin-chart-table/src/DataTable/DataTable.tsx:63]: Could this just be `showNextButton`?

File: plugins/plugin-chart-table/src/controlPanel.tsx [hunk 11]
```diff
@@ -192,7 +192,39 @@ const config: ControlPanelConfig = {
...
-        ['row_limit'],
+        [
+          {
+            name: 'backend_pagination',
+            config: {
+              type: 'CheckboxControl',
+              label: t('Backend pagination'),
+              description: t('Enable backend pagination'),
+              default: false,
+            },
```
Comment (villebro) [inline plugins/plugin-chart-table/src/controlPanel.tsx:203]: The label should probably be "Server pagination" and the description "Enable server side pagination of results"

**Review by villebro — COMMENTED** [review]
Mostly minor comments, looking good. However, I'm having trouble getting this to work, but it appears to be related to linking working differently on `npm@7` (I believe I had this working a few days ago, but now I can't get it to work). @ktmud - is it working ok for you?

**Comment by ktmud** [comment]
@villebro I didn't test this PR, but with npm 7, you need to manually run `npm i --legacy-peer-deps` in the plugin package after running `npm link`.

**Comment by simcha90** [comment]
> Mostly minor comments, looking good. However, I'm having trouble getting this to work, but it appears to be related to linking working differently on `npm@7` (I believe I had this working a few days ago, but now I can't get it to work). @ktmud - is it working ok for you?

but it also requires: https://github.com/apache/superset/pull/13137 to pull out

**Comment by villebro** [comment]
Thanks @simcha90 - I had indeed linked from the wrong branch. One last comment, as this feature isn't quite feature ready yet (slightly misleading percent metrics, no totals, no page numbers), perhaps we should the pagination control behind the cross filtering FF? We could then remove it when we have all features implemented and are certain that this works robustly.

**Review by villebro — APPROVED** [review]

**Comment by nuevoleonkx** [comment]
On which version is available this feature? I'm testing on 1.1 but apparently is not covered here. In addition, what would be the best approach to upgrade from 0.36? TIA

**Pull request merged**